src/Nethermind/Nethermind.Evm/CodeAnalysis/IL/ILCompiler.cs

```diff
@@ -1267,7 +1268,226 @@ private static void EmitSegmentBody(Emit<ExecuteSegment> method, OpcodeInfo[] co
                     method.LoadLocalAddress(localSpan);
                     method.Call(typeof(Span<byte>).GetMethod(nameof(Span<byte>.CopyTo), [typeof(Span<byte>)]));
                     break;
+                case Instruction.LOG0:
+                    method.StackLoadPrevious(stack, head, 1);
+                    method.Call(Word.GetUInt256);
+                    method.StoreLocal(uint256A);
+                    method.StackLoadPrevious(stack, head, 2);
+                    method.Call(Word.GetUInt256);
+                    method.StoreLocal(uint256B);
+                    method.StackPop(head, evmExceptionLabels[EvmExceptionType.StackUnderflow], 2);
+
+                    // UpdateMemoryCost
+                    method.LoadArgument(0);
+                    method.LoadField(GetFieldInfo(typeof(ILEvmState), nameof(ILEvmState.Memory)));
+                    method.LoadLocalAddress(gasAvailable);
+                    method.LoadLocalAddress(uint256A);
+                    method.LoadLocalAddress(uint256B);
+                    method.Call(
+                        typeof(VirtualMachine<VirtualMachine.NotTracing>).GetMethod(
+                            nameof(VirtualMachine<VirtualMachine.NotTracing>.UpdateMemoryCost)
+                        )
+                    );
+                    method.BranchIfFalse(evmExceptionLabels[EvmExceptionType.OutOfGas]);
+
+                    // update gas
+                    using (var cost = method.DeclareLocal<long>())
+                    {
+                        method.LoadLocal(gasAvailable);
+                        method.LoadConstant( GasCostOf.Log);
+                        method.LoadLocal(uint256B); // length
+                        method.LoadField(GetFieldInfo(typeof(UInt256), nameof(UInt256.u0)));
+                        method.Convert<long>();
+                        method.LoadConstant(GasCostOf.LogData);
+                        method.Multiply();
+                        method.Add();
+                        method.Convert<long>();
+                        method.StoreLocal(cost);
+                        method.LoadLocal(cost);
+                        method.CompareGreaterThan(); //  gasCost > gasAvailable
+                        method.BranchIfFalse(evmExceptionLabels[EvmExceptionType.OutOfGas]);
```

Comment:
why not do ``gasAvailable  = gasAvailable - cost``, and then ``if gasAvailable < 0 goto outOfGas``? would probably remove the need to the cost Local variable

Reply:
I tried doing this :
        ```
    il.LoadLocal(gasAvailable);
            // Handeling for LOG* cases
            switch (topicsCount)
            {
                case 0:
                    il.LoadConstant(GasCostOf.Log);
                    break;
                case 1:
                    il.LoadConstant(GasCostOf.LogTopic + GasCostOf.Log);
                    break;
                case 2:
                    il.LoadConstant(2 * GasCostOf.LogTopic + GasCostOf.Log);
                    break;
                case 3:
                    il.LoadConstant(3 * GasCostOf.LogTopic + GasCostOf.Log);
                    break;
                case 4:
                    il.LoadConstant(4 * GasCostOf.LogTopic + GasCostOf.Log);
                    break;
            }
            il.Convert<long>();
            il.LoadLocal(uint256B); // length
            il.LoadField(GetFieldInfo(typeof(UInt256), nameof(UInt256.u0)));
            il.Convert<long>();
            il.LoadConstant(GasCostOf.LogData);
            il.Multiply();
            il.Add();
            il.Subtract(); // gasAvailable - gasCost
            il.Duplicate();
            il.LoadConstant(0);
            il.Convert<long>();
            il.CompareGreaterThan();// gasAvailable - gasCost > 0
            il.BranchIfFalse(exceptions[EvmExceptionType.OutOfGas]);
            il.StoreLocal(gasAvailable); // gasAvailable = gasAvailable - gasCost
```


but I get an  `The JIT compiler encountered invalid IL code or an internal limitation.` probably because I still need the cost calculation for updating gasAvailable but when branching to the exception I need the stack to be empty which is why I introduced that variable